### PR TITLE
removes use of sc_client::Client from sc_finality_grandpa

### PR DESCRIPTION
**Pull request by seunlanlege**
See #4452

Introduces a new trait `ClientBackend` to `sc_client_api`, for methods that interact with the backend.
eg `fn lock_import_and_run`

### CODE REVIEW

File: client/api/src/backend.rs [hunk 1]
```diff
@@ -169,6 +169,14 @@ pub trait BlockImportOperation<Block: BlockT> {
 	fn mark_head(&mut self, id: BlockId<Block>) -> sp_blockchain::Result<()>;
 }
 
+/// trait for performing operations on the backend
```
Comment (NikVolf) [inline client/api/src/backend.rs]:
```suggestion
/// Interface for performing operations on the backend
```

File: client/api/src/backend.rs [hunk 2]
```diff
@@ -169,6 +169,14 @@ pub trait BlockImportOperation<Block: BlockT> {
...
+pub trait ClientBackend<Block: BlockT, B: Backend<Block>> {
+	/// Lock the import lock, and run operations inside.
+	fn lock_import_and_run<R, Err, F>(&self, f: F) -> Result<R, Err> where
```
Comment (NikVolf) [inline client/api/src/backend.rs]: where part formatting should be aligned with the rest of the codebase

**Comment by seunlanlege** [comment]
Pinging @gnunicorn @andresilva

File: client/api/src/backend.rs [hunk 3]
```diff
@@ -169,6 +169,15 @@ pub trait BlockImportOperation<Block: BlockT> {
 	fn mark_head(&mut self, id: BlockId<Block>) -> sp_blockchain::Result<()>;
 }
 
+/// Interface for performing operations on the backend.
+pub trait ClientBackend<Block: BlockT, B: Backend<Block>> {
```
Comment (gnunicorn) [inline client/api/src/backend.rs]: The name of the trait and its only function don't really match up. May `LockedImportRun` would be a better naming?

Not super happy about that name but it is def cleaner, unless you intent to add other functionality to it.

```suggestion
pub trait LockedImportRun<Block: BlockT, B: Backend<Block>> {
```

File: client/finality-grandpa/src/lib.rs [hunk 4]
```diff
@@ -56,15 +56,15 @@ use futures::prelude::*;
 use futures::StreamExt;
 use log::{debug, info};
 use futures::channel::mpsc;
-use sc_client_api::{BlockchainEvents, CallExecutor, backend::{AuxStore, Backend}, ExecutionStrategy};
-use sp_blockchain::{HeaderBackend, Error as ClientError};
+use sc_client_api::{ClientBackend, BlockchainEvents, CallExecutor, backend::{AuxStore, Backend}, ExecutionStrategy, Finalizer, TransactionFor};
```
Comment (gnunicorn) [inline client/finality-grandpa/src/lib.rs]: please wrap. no lines longer than 100 chars, please.

File: client/finality-grandpa/src/lib.rs [hunk 5]
```diff
@@ -257,6 +257,28 @@ impl<B, E, Block: BlockT, RA> BlockStatus<Block> for Arc<Client<B, E, Block, RA>
 	}
 }
 
+/// A trait that includes all the client functionalities grandpa requires.
+/// Ideally this would be a trait alias, we're not there yet.
```
Comment (gnunicorn) [inline client/finality-grandpa/src/lib.rs]: do we have a remote github issue we could link to here?

File: client/finality-grandpa/src/tests.rs [hunk 6]
```diff
@@ -18,10 +18,7 @@
 
 use super::*;
 use environment::HasVoted;
-use sc_network_test::{
-	Block, Hash, TestNetFactory, BlockImportAdapter, Peer,
-	PeersClient, PassThroughVerifier,
-};
+use sc_network_test::{Block, Hash, TestNetFactory, BlockImportAdapter, Peer, PeersClient, PassThroughVerifier, PeersFullClient};
```
Comment (gnunicorn) [inline client/finality-grandpa/src/tests.rs]: please wrap ;)

**Comment by seunlanlege** [comment]
Failing tests pass on my machine, not sure what the problem is.

**Comment by gnunicorn** [comment]
@seunlanlege please don't force push to any branches, especially when they are raised as PRs – it prevents github from picking up the difference between and thus reviewers can't just use the "changes since you last viewed" feature but have to review the entire PR again. We are squash merging in the end anyways, o no need to rebase or similar in between...

**Comment by seunlanlege** [comment]
So sorry about that @gnunicorn, will merge from now on.

**Pull request merged**